packages/react-dom/src/events/plugins/SimpleEventPlugin.js

```diff
@@ -158,6 +158,13 @@ function extractEvents(
     EventInterface,
   );
 
+  if (domEventName === 'focusin') {
```

Comment:
Rather than doing this here, we can probably do this in the switch statement above, where we assign focus events. It avoids having to do this check on every other event type then.

Reply:
We create the event after the the switch from the EventInterface. I would have to create a separate event interface for focusin and focusout e.g.

```js
export const FocusFocusEventInterface = {
  ...FocusEventInterface,
  type: 'focus',
};
export const FocusBlurEventInterface = {
  ...FocusEventInterface,
  type: 'blur',
};
```
But so far every event interface was based on a standards event interface.

Cloning the `FocusEventInterface` in the the switch seems excessive to me i.e.
```js
case 'focusin':
  EventInterface = {
    ...FocusEventInterface,
    type: 'focus'
  };
  break;
```

I can't tell what's worse: repeating checks that we know are unnecessary or copying objects.

Reply:
One possible approach is to expose two separate interfaces from `SyntheticEvent`. These could be `FocusInEventInterface` and `FocusOutEventInterface`, which both have a pre-defined field for type.

```jsx
export const FocusEventInterface = {
  ...UIEventInterface,
  relatedTarget: 0,
};

export const FocusInEventInterface = {
  ...FocusEventInterface,
  type: 'focus',
}

export const FocusOutEventInterface = {
  ...FocusEventInterface,
  type: 'blur',
}
```

Then you could seperate the switch statement to reference those instead:

```jsx
case 'focusin':
      EventInterface = FocusInEventInterface;
      break
    case 'focusout':
      EventInterface = FocusOutEventInterface;
      break
    case 'beforeblur':
    case 'afterblur':
      EventInterface = FocusEventInterface;
      break;
```

Reply:
That'll do.

Reply:
Went with 

```js
export const FocusInEventInterface = {
  ...FocusEventInterface,
  type: function() {
    return 'focus';
  },
};

export const FocusOutEventInterface = {
  ...FocusEventInterface,
  type: function() {
    return 'blur';
  },
};
```
since `SyntheticEvent` expects getters int the `Interface` object. No fat-arrow to match the other interfaces.